cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/GetRevisionTaskThreaded.java

```diff
@@ -137,14 +142,36 @@ public boolean add(Future<DocumentRevsList> documentRevsListFuture) {
         // we make the request at construction time...
         for (final BulkGetRequest request : requests) {
 
-            completionService.submit(new Callable<DocumentRevsList>() {
+            submittedJobs.add(completionService.submit(new Callable<DocumentRevsList>() {
                 @Override
                 public DocumentRevsList call() throws Exception {
                     return new DocumentRevsList(GetRevisionTaskThreaded.this.sourceDb.getRevisions
                             (request.id, request.revs, request.atts_since,
                                     GetRevisionTaskThreaded.this.pullAttachmentsInline));
                 }
-            });
+            }));
+        }
+    }
+
+    void cancel() {
+        synchronized (submittedJobs) {
+            for (Future<DocumentRevsList> f : submittedJobs) {
+                // Don't interrupt if already running, we'll get the running ones from the completed
+                // queue.
+                f.cancel(false);
+            }
+            // Cancelling will line a whole bunch of tasks up for addition to the queue, so we need
+            // to keep draining it
+            try {
+                while (!submittedJobs.isEmpty()) {
+                    Future<DocumentRevsList> r = completionService.take();
```

Comment:
@tomblench I've moved the `synchronized` to only surround the cancellation loop, which should allow for `submittedJobs.remove()` to be called from the iterator as well during the cancellation which I think eliminates the possibility of a job going untracked (i.e. I think there was a case before where the iterator poll could have pulled a job from the completion queue, but then ended up waiting behind the cancellation held lock to do the remove, which would prevent the cancellation completing because the jobs list would have one element left).

Additionally I changed it to poll with a timeout as per the previous comment.